File: 백준(최대한 많은 문제를 접해보자!)/200-자료구조1/08 덱.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Deque<String> deque = new ArrayDeque<>();
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            String command = st.nextToken();
            String str;

            switch (command) {
                case "push_back":
                    str = st.nextToken();
                    deque.offerLast(str);
                    break;
                case "push_front":
                    str = st.nextToken();
                    deque.offerFirst(str);
                    break;
                case "pop_front":
                    str = deque.pollFirst();
                    sb.append(str != null ? str : -1).append("\n");
                    break;
                case "pop_back":
                    str = deque.pollLast();
                    sb.append(str != null ? str : -1).append("\n");
                    break;
                case "size":
                    sb.append(deque.size()).append("\n");
                    break;
                case "empty":
                    sb.append(deque.isEmpty() ? 1 : 0).append("\n");
                    break;
                case "front":
                    str = deque.peekFirst();
                    sb.append(str != null ? str : -1).append("\n");
                    break;
                case "back":
                    str = deque.peekLast();
                    sb.append(str != null ? str : -1).append("\n");
                    break;
            }
        }
        System.out.println(sb);
    }
}

//
//
// 다른 풀이
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static int[] que = new int[20001];
	static int head = 10000;
	static int tail = 10000;
	static int size = 0;

	public static void main(String[] args) throws Exception{

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		StringBuilder sb = new StringBuilder();

		for(int i = 0; i < N; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			String S = st.nextToken();

			switch(S) {
			case "push_front" :
				push_front(Integer.parseInt(st.nextToken()));
				break;
			case "push_back" : 
				push_back(Integer.parseInt(st.nextToken()));
				break;
			case "pop_front" :
				sb.append(pop_front()).append("\n");
				break;
			case "pop_back" : 
				sb.append(pop_back()).append("\n");
				break;
			case "size" :
				sb.append(size()).append("\n");
				break;
			case "empty" :
				sb.append(empty()).append("\n");
				break;
			case "front" :
				sb.append(front()).append("\n");
				break;
			case "back" :
				sb.append(back()).append("\n");
				break;
			}
		}
		System.out.println(sb);
	}

	private static int pop_back() {
		if(size == 0) {
			return -1;
		}else {
			int P = que[tail];
			tail--;
			size--;
			return P;
		}		
	}

	private static int pop_front() {
		if(size == 0) {
			return -1;
		}else {
			int P = que[head + 1];
			head++;
			size--;
			return P;
		}
	}

	private static void push_back(int parseInt) {
		tail++;
		size++;
		que[tail] = parseInt;
		
	}

	private static void push_front(int parseInt) {
		que[head] = parseInt;
		head--;
		size++;
		
	}

	private static int back() {
		if(size == 0) {
			return -1;
		}else {
			return que[tail];
		}
	}

	private static int front() {
		if(size == 0) {
			return -1;
		}else {
			return que[head+1];
		}
	}

	private static int empty() {
		if(size  == 0) {
			return 1;
		}else {
			return 0;
		}
	}

	private static int size() {
		return size;
	}
}

File: 백준(최대한 많은 문제를 접해보자!)/200-자료구조1/08 덱.js
let fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
let input = fs.readFileSync(filePath).toString().trim().split('\n');

input.shift();

input = input.map((v) => v.trim().split(' '));

let deque = [];
let answer = [];

for (let command of input) {
  if (command[0] === 'push_front') deque.unshift(command[1]);
  if (command[0] === 'push_back') deque.push(command[1]);
  if (command[0] === 'pop_front') {
    if (deque.length) answer.push(deque.shift());
    else answer.push(-1);
  }
  if (command[0] === 'pop_back') {
    if (deque.length) answer.push(deque.pop());
    else answer.push(-1);
  }
  if (command[0] === 'size') answer.push(deque.length);
  if (command[0] === 'empty') {
    if (deque.length) answer.push(0);
    else answer.push(1);
  }
  if (command[0] === 'front') {
    if (deque.length) answer.push(deque[0]);
    else answer.push(-1);
  }
  if (command[0] === 'back') {
    if (deque.length) answer.push(deque[deque.length - 1]);
    else answer.push(-1);
  }
}

console.log(answer.join('\n'));

//
//
// 다른 풀이
const I = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const [N, deque, O] = [+I[0], [], []];
for (let i = 1; i <= N; i++) {
  const [command, num] = I[i].split(' ');
  const l = deque.length;
  switch (command) {
    case 'push_back':
      deque.push(num);
      break;
    case 'push_front':
      deque.unshift(num);
      break;
    case 'pop_back':
      O.push(l ? deque.pop() : -1);
      break;
    case 'pop_front':
      O.push(l ? deque.shift() : -1);
      break;
    case 'size':
      O.push(l);
      break;
    case 'empty':
      O.push(+!l);
      break;
    case 'front':
      O.push(l ? deque[0] : -1);
      break;
    case 'back':
      O.push(l ? deque[l - 1] : -1);
      break;
  }
}
console.log(O.join('\n'));
